js_modules/dagster-ui/packages/ui-core/src/runs/useRunsForTimeline.tsx

```diff
@@ -23,52 +24,180 @@ import {repoAddressAsHumanString} from '../workspace/repoAddressAsString';
+  }, [filter]);
   const [start, end] = range;
 
   const startSec = start / 1000.0;
   const endSec = end / 1000.0;
 
-  const unteriminatedRunsQueryData = useQuery<
-    UnterminatedRunTimelineQuery,
-    UnterminatedRunTimelineQueryVariables
-  >(UNTERMINATED_RUN_TIMELINE_QUERY, {
-    notifyOnNetworkStatusChange: true,
-    // With a very large number of runs, operating on the Apollo cache is too expensive and
-    // can block the main thread. This data has to be up-to-the-second fresh anyway, so just
-    // skip the cache entirely.
-    fetchPolicy: 'no-cache',
-    variables: {
-      inProgressFilter: {
-        ...runsFilter,
-        statuses: [RunStatus.CANCELING, RunStatus.STARTED],
-        createdBefore: endSec,
-      },
-    },
-  });
+  const buckets = useMemo(() => {
+    const buckets = [];
+    for (let time = start; time < end; time += BUCKET_SIZE) {
+      buckets.push([time, Math.min(end, time + BUCKET_SIZE)] as const);
+    }
 
-  const terminatedRunsQueryData = useQuery<
-    TerimatedRunTimelineQuery,
-    TerimatedRunTimelineQueryVariables
-  >(TERMINATED_RUN_TIMELINE_QUERY, {
-    notifyOnNetworkStatusChange: true,
-    // With a very large number of runs, operating on the Apollo cache is too expensive and
-    // can block the main thread. This data has to be up-to-the-second fresh anyway, so just
-    // skip the cache entirely.
-    fetchPolicy: 'no-cache',
-    variables: {
-      terminatedFilter: {
-        ...runsFilter,
-        statuses: Array.from(doneStatuses),
-        createdBefore: endSec,
-        updatedAfter: startSec,
-      },
-    },
+    return buckets;
+  }, [start, end]);
+
+  const client = useApolloClient();
+
+  const [terminatedRunsQueryData, setTerminatedRunsData] = useState<{
+    data: RunTimelineFragment[] | undefined;
+    loading: boolean;
+    error: any;
+    called: boolean;
+  }>({
+    data: undefined,
+    loading: true,
+    error: undefined,
+    called: false,
+  });
+  const [unterminatedRunsQueryData, setUnterminatedRunsData] = useState<{
+    data: RunTimelineFragment[] | undefined;
+    loading: boolean;
+    error: any;
+    called: boolean;
+  }>({
+    data: undefined,
+    loading: true,
+    error: undefined,
+    called: false,
   });
+  const {data: unterminatedRunsData, loading: loadingUnterminatedRunsData} =
+    unterminatedRunsQueryData;
+  const {data: terminatedRunsData, loading: loadingTerminatedRunsData} = terminatedRunsQueryData;
+
+  const fetchUnterminatedRunsQueryData = useCallback(async () => {
+    setUnterminatedRunsData(({data}) => ({
+      data,
+      loading: true,
+      called: true,
+      error: undefined,
+    }));
+    let cursor: string | undefined = undefined;
+    const results = await Promise.all(
+      buckets.map(
+        ([start, end]) =>
+          new Promise<RunTimelineFragment[]>(async (res) => {
+            let hasMoreData = true;
+            const dataSoFar: RunTimelineFragment[] = [];
+            while (hasMoreData) {
```

Comment:
I don't find it very legible having this function with non trivial control flow inlined like this

Reply:
Yeah I'm going to refactor this